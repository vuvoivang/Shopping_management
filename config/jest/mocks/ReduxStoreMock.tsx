import React from 'react';
import configureMockStore from 'redux-mock-store';
import { merge } from 'lodash';
import { Provider } from 'react-redux';
import thunkMiddleware from 'redux-thunk';
import { shallow, mount } from 'enzyme';
import { routerMiddleware } from 'connected-react-router';
import { AppConstant } from '../../../src/constants/app.constant';
import { getAppState } from '../TestHelper';
import { defaultLanguage } from './IntlHelper';
import history from '../../../src/helpers/history.helper';
import { Router } from 'react-router';
import { getLanguageMessages } from '../../../src/utilities/i18n/i18n.utility';
import { IntlProvider } from 'react-intl';

const languageInitialState = {
  [AppConstant.redux.LANGUAGE_STATE]: {
    locale: 'en',
    formattedMessages: {
      en: defaultLanguage
    }
  }
};

export const defaultStore = {
  router: {
    location: {}
  }
};
const { locale, languageMessages } = getLanguageMessages(navigator.language.split(/[-_]/)[0]);

export const shallowWithProvider = children => (store = getMockStore({})) => shallow(<Provider
  store={store}>{children}</Provider>);
export const mountWithProvider = children => (store = getMockStore({})) => mount(<Provider
  store={store}>{children}</Provider>);

export const mountWithProviderAndRouter = children => (store = getMockStore({})) => {
  return mount(
    (
      <Provider store={store}>
        <IntlProvider locale={locale} messages={languageMessages}>
          <Router history={history}>
            {children}
          </Router>
        </IntlProvider>
      </Provider>
    )
  );
}

export const mountWithProviderAndStore = (
  children,
  userType?: string,
  store?: { [key: string]: any },
  middleware = []
) => {
  let mockStore;
  if (!store) {
    mockStore = getMockStore(defaultStore, middleware);
  } else {
    mockStore = getMockStore(store, middleware);
  }
  return mountWithProvider(children)(mockStore);
};

export const getMockStore = (state, middleware = []) => {
  const middlewares = [thunkMiddleware, routerMiddleware(history)].concat(middleware);
  const mockStore = configureMockStore(middlewares);
  const languageAddedState = merge({}, languageInitialState, getAppState(), state);
  return mockStore(languageAddedState);
};

export const getMockStoreWithConfig = ({ state = {}, middleware = [], epic, epics = [] }) => {
  let reduxMiddlewares = [];
  if (middleware && middleware.length) {
    reduxMiddlewares.push(...middleware);
  }

  return getMockStore(state, reduxMiddlewares);
};
