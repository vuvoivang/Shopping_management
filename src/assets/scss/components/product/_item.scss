.product-item {
  margin: 20px;
  &__detail {
    @include button(#fff, $color-pr-1, #fff, $color-pr-2);
  }
}
