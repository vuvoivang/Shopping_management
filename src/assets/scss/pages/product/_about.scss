.c-about {
  &__text {
    padding: 25px 0;
    &--primary {
      color: $color-pr-1;
    }
  }
  &__desc {
    text-align: center;
    letter-spacing: 1px;
    line-height: 1.5rem;
  }
}
@media only screen and (max-width: 500px) {
  .c-about {
    padding: 20px 30px;
    &__text {
      font-size: 22px;
      padding: 20px 0;
    }
    &__desc {
      font-size: 15px;
    }
  }
}
