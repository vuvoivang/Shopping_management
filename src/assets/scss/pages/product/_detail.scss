.c-product-detail {
  &__card-item{
    border-radius: 20px;
    overflow: hidden;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    padding: 100px 200px;
  }
  &__info {
    @include flex(column, space-around, flex-start, wrap);
    padding: $space;
    padding-left: $space-2x;
    max-width:600px;
    h1 {
      font-size: $fs-xl + 3px;
    }
    p {
      margin: 30px 0;
      font-size: $fs-sm + 2px;
      line-height: 2rem;
    }
  }
  &__button-add-to-cart {
    @include button(#fff, $color-red, #fff, $color-orange);
    margin-right: $space;
    padding: 15px 30px;
    margin-right: 7px;
  }
  &__button-download-desc {
    @include button(#fff, $color-pr-1, #fff, $color-pr-2);
  }
  &__card-price {
    margin-bottom: $space-2x;
    color: $color-red;
    font-size: $fs-md + 3px;
    font-weight: $fw-bold-1;
    text-align: center;
  }
}
.l_wrapper {
  @include flex(row, space-around, center, wrap);
}
