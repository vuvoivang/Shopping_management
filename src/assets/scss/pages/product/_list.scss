.c-product-list {
  @include flex(column, center, center, wrap);
  padding: 80px;
  &__row {
    width: 100%;
    @include flex(row, center, center, wrap);
  }
  &__header {
    color: $color-orange;
    margin-bottom: $space-2x;
  }
}
