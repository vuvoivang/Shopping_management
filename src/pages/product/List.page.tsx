import ProductItem from '../../components/product/ProductItem.component';

const ProductList: React.FC = () => {
  let productList = [
    { id: 1, name: 'A', price: 10000 },
    { id: 2, name: 'B', price: 15000 },
    { id: 3, name: 'C', price: 12000 }
  ];
  return (
    <div className="product-list">
      <h1>Product List</h1>
      <div className="product-list__list">
        {productList.map(item => (
          <div key={item.id}>
            <ProductItem id={item.id} name={item.name} price={item.price} />
          </div>
        ))}
      </div>
    </div>
  );
};

export default ProductList;
