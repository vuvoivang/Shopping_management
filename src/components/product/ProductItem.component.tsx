import { Product } from 'models/product.model';
import { NavLink } from 'react-router-dom';

interface ProductItemProps {
  product: Product;
}

const ProductItem: React.FC<ProductItemProps> = ({ product: { id, name, price, image, detail } }: ProductItemProps) => {
  const addToCart = () => {
    console.log('add to cart');
  };
  return (
    <div className="c-product-item">
      <div className="c-product-item__card-item">
        <div>
          <img className="c-product-item__card-image" style={{ objectFit: 'contain' }} src={image} alt="Smart watch" height="250" />
          <div className="c-product-item__card-content">
            <div className="c-product-item__card-info">
              <h4>{name}</h4>
              <p>{detail}</p>
              <div className="c-product-item__card-price">{`${price}.000 VND`}</div>
            </div>
            <div className="c-product-item__action">
              <NavLink to={`/products/${id}`}>
                <button type="button" className="c-product-item__add-to-cart">
                  Add to cart
                </button>
              </NavLink>
              <button type="button" className="c-product-item__detail" onClick={addToCart}>
                Detail
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};
export default ProductItem;
