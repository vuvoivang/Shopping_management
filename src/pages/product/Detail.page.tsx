import { useState, useEffect } from 'react';
import { Link, useParams } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faPlusCircle, faDownload } from '@fortawesome/free-solid-svg-icons';
import { fetchProductDetail } from '../../assets/dummy-data/product.data';

interface ParamTypes {
  productId: string;
}

const ProductDetail: React.FC = () => {
  const [product, setProduct] = useState<Product>();
  const { productId } = useParams<ParamTypes>();
  const fetchData = async id => {
    let productDetail = await fetchProductDetail(id);
    setProduct(productDetail);
  };
  useEffect(() => {
    try {
      fetchData((productId as unknown) as string);
    } catch (err) {
      if (typeof err === 'string') {
        // eslint-disable-next-line no-alert
        alert(err.toUpperCase());
      } else if (err instanceof Error) {
        // eslint-disable-next-line no-alert
        alert(err.message);
      }
    }
  }, []);
  return (
    <div className="c-product-detail">
      {product && (
        <div
          style={{
            borderRadius: '20px',
            overflow: 'hidden',
            display: 'flex',
            flexDirection: 'column',
            justifyContent: 'center',
            alignItems: 'center',
            padding: '100px 200px'
          }}
        >
          <div className="l_wrapper d-flex justify-content-around align-items-center">
            <div className="c-product-detail__image">
              <img alt="smart watch" src={product.image} height="300" style={{ objectFit: 'contain' }} />
            </div>
            <div className="c-product-detail__info" style={{ maxWidth: 600 }}>
              <h1 style={{ color: 'blue' }}>{product.name}</h1>
              <p>{product.detail}</p>
              <div className="c-product-detail__card-price">{`${product.price}.000 VND`}</div>
              <div className="btn-group">
                <Link to="/products">
                  <button type="button" className="c-product-detail__button-add-to-cart" style={{ padding: '15px 30px' }}>
                    <FontAwesomeIcon icon={faPlusCircle} size="lg" style={{ marginRight: 7 }} /> Add to cart
                  </button>
                </Link>
                <button type="button" className="c-product-detail__button-download-desc" style={{ padding: '15px 30px' }}>
                  <a href={Number(product.id) % 2 === 0 ? '/files/EvenProduct.xlsx' : '/files/OddProduct.xlsx'} target="_self">
                    <FontAwesomeIcon icon={faDownload} size="lg" style={{ marginRight: 7 }} /> Down description file
                  </a>
                </button>
              </div>
            </div>
          </div>
        </div>
      )}
    </div>
  );
};

export default ProductDetail;
