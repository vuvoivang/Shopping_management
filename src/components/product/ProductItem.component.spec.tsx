import React from 'react';
import { Product } from 'models/product.model';
import { mountWithProviderAndRouter } from '../../../config/jest/mocks/ReduxStoreMock';
import ProductItem from './ProductItem.component';


describe('<ProductItem />', () => {
  const sampleProduct: Product = {
    id: 1,
    name: 'Hero Watch',
    detail: 'Apple Watch is the ultimate device for a healthy life. Available in three models: Apple Watch',
    price: 90,
    hero: 'OMG This just came out today!',
    image: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQAzV2p08eELUS-trvsvKOG1YLMTtwVTD4z3Z4BozUTilSEhIJ1MLaY7ADGurmT0E6yZfk&usqp=CAU'
  };

  const setUp = () => {
    const wrapper = mountWithProviderAndRouter(<ProductItem product={sampleProduct} />)();
    return { wrapper };
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('render as expect', () => {
    const { wrapper } = setUp();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.c-product-item__card-image').exists()).toBeTruthy();
    expect(wrapper.find('.c-product-item__card-content').exists()).toBeTruthy();
    expect(
      wrapper
        .find('.c-product-item__card-info')
        .find('h4')
        .text()
    ).toEqual(sampleProduct.name);
    expect(
      wrapper
        .find('.c-product-item__card-info')
        .find('p')
        .text()
    ).toEqual(sampleProduct.detail);
    expect(wrapper.find('.c-product-item__card-price').text()).toEqual(`${sampleProduct.price}.000 VND`);
  });

  it('NavLink path to detail', () => {
    const { wrapper } = setUp();
    const NavLink = wrapper.find('NavLink');
    expect(NavLink).toHaveLength(1);
    expect(NavLink.prop('to')).toEqual(`/products/${sampleProduct.id}`);
  });
});
