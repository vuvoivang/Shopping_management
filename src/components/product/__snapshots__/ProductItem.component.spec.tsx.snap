// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductItem /> render as expect 1`] = `ReactWrapper {}`;
